services/intersection-api/api/src/main/java/us/dot/its/jpo/ode/api/asn1/BsmDecoder.java

```diff
@@ -0,0 +1,129 @@
+            OdeBsmData bsm = getAsOdeJson(decodedXml);
+
+            // build output data structure
+            DecodedMessage decodedMessage = new BsmDecodedMessage(bsm, message.getAsn1Message(), "");
+            return decodedMessage;
+            
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+            return new BsmDecodedMessage(null, message.getAsn1Message(), e.getMessage());
+        } catch (Exception e) {
+            e.printStackTrace();
+            return new BsmDecodedMessage(null, message.getAsn1Message(), e.getMessage());
+        }
+    }
+
+    @Override
+    public OdeData getAsOdeData(String encodedData) {
+        OdeMsgPayload payload = new OdeAsn1Payload(new OdeHexByteArray(encodedData));
+
+        // construct metadata
+        OdeBsmMetadata metadata = new OdeBsmMetadata(payload);
+        metadata.setOdeReceivedAt(DecoderManager.getOdeReceivedAt());
+        metadata.setRecordType(RecordType.bsmTx);
+        metadata.setSecurityResultCode(SecurityResultCode.success);
+
+        // construct metadata: receivedMessageDetails
+        ReceivedMessageDetails receivedMessageDetails = new ReceivedMessageDetails();
+        receivedMessageDetails.setRxSource(RxSource.RV);
+
+        // construct metadata: locationData
+        OdeLogMsgMetadataLocation locationData = new OdeLogMsgMetadataLocation();
+        receivedMessageDetails.setLocationData(locationData);
+
+        metadata.setReceivedMessageDetails(receivedMessageDetails);
+        metadata.setBsmSource(BsmSource.RV);
+
+        Asn1Encoding unsecuredDataEncoding = new Asn1Encoding("unsecuredData", "MessageFrame",
+                EncodingRule.UPER);
+        metadata.addEncoding(unsecuredDataEncoding);
+
+        // construct odeData
+        return new OdeAsn1Data(metadata, payload);
+    }
+
+    @Override
+    public OdeBsmData getAsOdeJson(String consumedData) throws XmlUtilsException {
+        ObjectNode consumed = XmlUtils.toObjectNode(consumedData);
+
+        JsonNode metadataNode = consumed.findValue(AppContext.METADATA_STRING);
+        if (metadataNode instanceof ObjectNode) {
+            ObjectNode object = (ObjectNode) metadataNode;
+            object.remove(AppContext.ENCODINGS_STRING);
+        }
+        
+        OdeBsmMetadata metadata = (OdeBsmMetadata) JsonUtils.fromJson(
+            metadataNode.toString(), OdeBsmMetadata.class);
+
+        /*
+        *  ODE-755 and ODE-765 Starting with schemaVersion=5 receivedMessageDetails 
+        *  will be present in BSM metadata. None should be present in prior versions.
+        */
+        if (metadata.getSchemaVersion() <= 4) {
+            metadata.setReceivedMessageDetails(null);
+        }
+        
+        OdeBsmPayload payload = new OdeBsmPayload(
+            BsmBuilder.genericBsm(consumed.findValue("BasicSafetyMessage")));
+        return new OdeBsmData(metadata, payload);
+    }
+
+    public String mockDecodeXmlWithAcm(String xmlMessage){
```

Comment:
For these decoder classes under asn1, should this mocking be left to the tests rather than the class?

Reply:
this was unused, removed